README
This repo contains the following: Magento 2.2.1, React UI and Hapi.js REST proxy between react and magento. **This is not a react theme for magento**, but a complete replacement of the frontend. Idea is that the magento is running in the background as headless system, so customers never really interact directly with magento (except for the store owner who can access the admin interface).

## Vagrant environment

If you want to try this system out, I have provided vagrant scripts that create ready to use environment.

Before starting up environment, **replace** empty file "Magento-CE-2.2.1_sample_data-2017-11-04-12-09-15.tar.bz2" in VagrantScripts-folder with a proper downloaded copy of magento with sampledata. This can't be done automatically, thanks to some braindead policy of keeping downloads behind login.

Vagrantfile assumes you are using vmware fusion (osx) provider, make necessary changes if using something else like virtualbox. To try out the environment, start up virtual machine with command **vagrant up** . This takes some time, so reserve lots of coffee :)

After the machine has been brought up, you can access magento admin via url:

http://localhost:3090/admin_abc1

Username is: **admin** and password is: **pass1234**

Current script doesn't start up node.js proxy, so you must manually ssh to the machine and start it up like so:

* vagrant ssh
* cd /vagrant/proxy
* node .

Leave node running on the background, and now you should (tm) be able to access react app with url: http://localhost:3080/

If you want to start development server for frontend, use command "npm start" in folder '/vagrant/frontend'.

## Notes

**npm start** will fail, if you have installed node modules outside vagrant environment and vice versa. That's why we remove node_modules folder on init.


Changelog.md
## [10/2020]:
* Switched back to MariaDB (from MySQL 8)
* Updated Magento from 2.2.8 to 2.2.11
* Replaced vagrant with docker
* Rewrote the proxy module with typescript

## [01/2019]:
* Replaced MariaDB & ArangoDB with MySQL 8
* Updated Node.js from 8 to 10
* Updated magento to 2.2.7
* Updated PHP from 7.0 to 7.1


VagrantScripts/base.sh
#!/usr/bin/env bash

# disable selinux for current boot
setenforce 0

# disable selinux permanently
sed -i 's/SELINUX=enforcing/SELINUX=disabled/' /etc/sysconfig/selinux
sed -i 's/SELINUX=enforcing/SELINUX=disabled/' /etc/selinux/config

yum install -q -y epel-release

# Install mysql 8 , elastic and node.js 10 repos
cp /vagrant/VagrantScripts/elastic.repo /etc/yum.repos.d/
yum install -y https://dev.mysql.com/get/mysql80-community-release-el7-1.noarch.rpm
curl --silent --location https://rpm.nodesource.com/setup_10.x | sudo bash -

# Enable installation after epel is installed
yum install -q -y ntp vim-enhanced wget git nodejs mysql-community-server mysql-shell nginx tree elasticsearch java-1.8.0-openjdk-headless

# Set the correct time
ntpdate -u pool.ntp.org

# Replace mysql configuration
cp /vagrant/VagrantScripts/my.cnf /etc/

# Enable & start ntp/mariadb
systemctl enable ntpd
systemctl enable mysqld
systemctl start ntpd
systemctl start mysqld

# PHP 7.1.x install:
PHP_VERSION="71"
yum install -q -y http://rpms.remirepo.net/enterprise/remi-release-7.rpm
yum install -q -y \
  php${PHP_VERSION}-php-cli \
  php${PHP_VERSION}-php-fpm \
  php${PHP_VERSION}-php-mysqli \
  php${PHP_VERSION}-php-mbstring \
  php${PHP_VERSION}-php-gd \
  php${PHP_VERSION}-php-mcrypt \
  php${PHP_VERSION}-php-zip \
  php${PHP_VERSION}-php-soap \
  php${PHP_VERSION}-php-opcache \
  php${PHP_VERSION}-php-intl \
  php${PHP_VERSION}-php-bcmath \
  php${PHP_VERSION}-php-xml

ln -s /usr/bin/php${PHP_VERSION} /usr/bin/php

# Copy config files
cp /vagrant/VagrantScripts/nginx.conf /etc/nginx/nginx.conf
cp /vagrant/VagrantScripts/www.conf /etc/opt/remi/php71/php-fpm.d/www.conf

# Setup more sane dev environment (vim + git settings)
sudo -u vagrant cp /vagrant/VagrantScripts/vimrc /home/vagrant/.vimrc
sudo -u vagrant git config --global color.ui auto

#
chgrp vagrant /var/opt/remi/php71/lib/php/session
chgrp vagrant /var/opt/remi/php71/lib/php/wsdlcache
chgrp vagrant /var/opt/remi/php71/lib/php/opcache
chown vagrant /var/opt/remi/php71/log/php-fpm

# Increase php memory limit
sed -i 's/memory_limit = 128M/memory_limit = 770M/' /etc/opt/remi/php71/php.ini

# Create directory and database for magento + proxy
mkdir -p /services/magento
cp -R /vagrant/VagrantScripts/tools /services
chown vagrant:vagrant -R /services

# Change temp password to: Mysql#1234
PASS=`grep "A temporary password is generated for" /var/log/mysqld.log | tail -1 | rev | cut -d" " -f 1 | rev`
mysqladmin -u root -p"$PASS" password Mysql#1234
echo "MySQL root password is: Mysql#1234"
PASS="Mysql#1234"

MYSQL_PWD="$PASS" mysqladmin -u root create magento2
MYSQL_PWD="$PASS" mysqladmin -u root create magentoproxy
MYSQL_PWD="$PASS" mysql -u root -e "create user 'magento'@'localhost' identified by 'Magentopass_1234'"
MYSQL_PWD="$PASS" mysql -u root -e "grant all privileges on magento2.* to 'magento'@'localhost'"
MYSQL_PWD="$PASS" mysql -u root -e "create user 'mageproxy'@'localhost' identified by 'Proxypass_1234'"
MYSQL_PWD="$PASS" mysql -u root -e "grant all privileges on magentoproxy.* to 'mageproxy'@'localhost'"
MYSQL_PWD="$PASS" mysqladmin -u root flush-privileges


# Reinstall npm packages
cd /vagrant/frontend ; rm -rf node_modules
sudo -u vagrant npm ci
sudo -u vagrant npm run build

cd ../proxy ; rm -rf node_modules
sudo -u vagrant npm ci

# Reduce elasticsearch memory usage
sed -ie 's/Xms2g/Xms768m/' /etc/elasticsearch/jvm.options
sed -ie 's/Xmx2g/Xmx768m/' /etc/elasticsearch/jvm.options

# Enable more services
systemctl enable nginx
systemctl enable php71-php-fpm
systemctl enable elasticsearch

systemctl start nginx
systemctl start php71-php-fpm
systemctl start elasticsearch

# Install magento
cd /services/magento
sudo -u vagrant tar xf /vagrant/VagrantScripts/Magento-CE-2.2.8_sample_data-2019-03-18-09-16-57.tar.bz2
sudo -u vagrant php bin/magento setup:install \
    --db-user=magento --db-password=Magentopass_1234 --db-name=magento2 \
    --admin-firstname=vagrant --admin-lastname=vagrant --admin-user=admin\
    --admin-password=pass1234 --admin-email="changeme@mailinator.com" --backend-frontname="admin_abc1"\
    --base-url=http://localhost:3090/

# Install composer
cp /vagrant/VagrantScripts/composer.phar /usr/local/bin/composer

# Disable magentos caches and enable developer mode
#sudo -u vagrant php bin/magento cache:disable
#sudo -u vagrant php bin/magento deploy:mode:set developer

# Install ElasticIndexer module
cd /services/magento

#sudo -u vagrant /usr/local/bin/composer config repositories.0 vcs https://github.com/mcfizh/elasticindexer
#sudo -u vagrant /usr/local/bin/composer require --prefer-source "mcfish/elasticindexer"

#sudo -u vagrant php bin/magento setup:upgrade
#sudo -u vagrant php bin/magento setup:di:compile

# Create magento integration & change config values
cd /services/tools
sudo -u vagrant php integration.php
sudo -u vagrant php changeConfigs.php

# Flush cache
cd /services/magento
sudo -u vagrant php bin/magento cache:flush

# Create cron job for vagrant user
sudo -u vagrant crontab /vagrant/VagrantScripts/cron
